[SPARK-21195][CORE] MetricSystem should pick up dynamically registered metrics in sources

robert3005 opened the pull request:
### What changes were proposed in this pull request?
MetricSystem picks up new metrics from sources that are added throughout execution. If you do measurements via dynamic proxies you might not want to redeclare all metrics that the proxies will create and you'd prefer them to get populated as they're being produced. Right now all sources are processed only onceat startup and metrics are picked up only if they have been registered statically at compile time. Behaviour I am proposing lets you not have to declare metrics in two places.

This had been previously suggested in https://github.com/apache/spark/pull/18406 and https://github.com/apache/spark/pull/29980. I have reduced the scope of the change to just dynamic metric registration.

### Why are the changes needed?
Currently there's no way to access MetricRegistry that MetricsSystem uses to hold its state and as such it's not possible to reprocess a source. MetricsSystem throws if any metric had already been registered previously.

n.b. the MetricRegistry is added as a constructor argument to make testing easier but could as well be accessed via reflection as a private variable.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added tests

Commit:
Merge existing registry with default one or configure default metric registry

Commit:
variable name

Commit:
better variable  name

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented:
ok to test

SparkQA commented:
**[Test build #135325 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135325/testReport)** for PR 31267 at commit [`56706ff`](https://github.com/apache/spark/commit/56706ff0a6643d3fc22380b83b3a6dec8bf72f8c).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39905/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39905/

SparkQA commented:
**[Test build #135325 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135325/testReport)** for PR 31267 at commit [`56706ff`](https://github.com/apache/spark/commit/56706ff0a6643d3fc22380b83b3a6dec8bf72f8c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39905/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135325/

robert3005 commented:
Looks like jenkins tests pass fine andn giithub actions flaked

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #137423 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137423/testReport)** for PR 31267 at commit [`c808690`](https://github.com/apache/spark/commit/c8086901097ff468a80f648b5601e824a2138fbb).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42000/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41999/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/41999/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42000/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42000/

SparkQA commented:
**[Test build #137423 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137423/testReport)** for PR 31267 at commit [`c808690`](https://github.com/apache/spark/commit/c8086901097ff468a80f648b5601e824a2138fbb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/137423/

mridulm commented on core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala:
```diff
@@ -152,28 +155,25 @@ private[spark] class MetricsSystem private (
+  def removeSource(source: Source): Unit = sourcesWithListeners.synchronized {
     val regName = buildRegistryName(source)
     registry.removeMatching((name: String, _: Metric) => name.startsWith(regName))
+    sourcesWithListeners.get(source).foreach(source.metricRegistry.removeListener)
+    sourcesWithListeners.remove(source)
```
```suggestion
sourcesWithListeners.remove(source).foreach(source.metricRegistry.removeListener)
```

mridulm commented on core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala:
```diff
@@ -108,6 +110,7 @@ private[spark] class MetricsSystem private (
     if (running) {
       sinks.foreach(_.stop)
       registry.removeMatching((_: String, _: Metric) => true)
+      sourcesWithListeners.keySet.foreach(removeSource)
```
synchronize on `sourcesWithListeners`

mridulm commented on core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala:
```diff
@@ -152,28 +155,25 @@ private[spark] class MetricsSystem private (
+      val listener = new MetricsSystemListener(buildRegistryName(source))
+      source.metricRegistry.addListener(listener)
+      sourcesWithListeners += source -> listener
     } catch {
```
Let us keep the scope of the lock narrow to match what was there earlier, unless there is a good reason to expand the critical section ?

mridulm commented on core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala:
```diff
@@ -242,7 +277,14 @@ private[spark] object MetricsSystem {
+     instance: String,
+     conf: SparkConf,
+     registry: MetricRegistry): MetricsSystem = {
+    new MetricsSystem(instance, conf, registry)
```
Rely on default registry param if unspecified ?

mridulm reviewed (commented):
Thanks for working on this @robert3005, took a quick pass through it.

mridulm commented:
+CC @HyukjinKwon, @jerryshao who have worked on this in past.

HyukjinKwon commented:
Thanks for cc'ing me @mridulm.

Actually I reviewed this. It makes sense but I am not very used to this code path to be honest ...

mridulm commented:
Thanks @HyukjinKwon :-)
+CC @srowen

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #137635 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137635/testReport)** for PR 31267 at commit [`c808690`](https://github.com/apache/spark/commit/c8086901097ff468a80f648b5601e824a2138fbb).

SparkQA commented:
**[Test build #137635 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137635/testReport)** for PR 31267 at commit [`c808690`](https://github.com/apache/spark/commit/c8086901097ff468a80f648b5601e824a2138fbb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/137635/

robert3005 commented:
This pr shouldn't be stale. I forgot about but will address comments eow

robert3005 committed:
Update core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala

Co-authored-by: Mridul Muralidharan <1591700+mridulm@users.noreply.github.com>

Commit:
default

Commit:
less

Commit:
address

robert3005 commented:
Addressed the comments. While reviewing the code I noticed there was a small thing that I missed - the error on duplicate metric name will happen now in the listener so I kept the logic that suppressed that error when registering

SparkQA commented:
**[Test build #142115 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142115/testReport)** for PR 31267 at commit [`2e6f29b`](https://github.com/apache/spark/commit/2e6f29baa2938a8473020d59351f6c40eaf9f624).

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46627/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46627/

SparkQA commented:
**[Test build #142115 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142115/testReport)** for PR 31267 at commit [`2e6f29b`](https://github.com/apache/spark/commit/2e6f29baa2938a8473020d59351f6c40eaf9f624).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142115/

Commit:
boun

SparkQA commented:
**[Test build #142174 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142174/testReport)** for PR 31267 at commit [`e375448`](https://github.com/apache/spark/commit/e3754487a8fc49622059a47e403945f2850731f2).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46686/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46686/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46686/

SparkQA commented:
**[Test build #142174 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142174/testReport)** for PR 31267 at commit [`e375448`](https://github.com/apache/spark/commit/e3754487a8fc49622059a47e403945f2850731f2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142174/

SparkQA commented:
**[Test build #143664 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143664/testReport)** for PR 31267 at commit [`e375448`](https://github.com/apache/spark/commit/e3754487a8fc49622059a47e403945f2850731f2).

SparkQA commented:
**[Test build #143664 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143664/testReport)** for PR 31267 at commit [`e375448`](https://github.com/apache/spark/commit/e3754487a8fc49622059a47e403945f2850731f2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143664/
